src/bit-systems/audio-zone-system.ts

```diff
@@ -0,0 +1,318 @@
+      applySourceParams(
+        source,
+        Object.assign(
+          {},
+          inOutParams,
+          outInParams,
+          paramsReducer(
+            {
+              gain: outInParams?.gain
+            },
+            {
+              gain: inOutParams?.gain
+            }
+          )
+        )
+      );
+    }
+  };
+})();
+
+const REDUCED_KEYS = [
+  "gain",
+  "maxDistance",
+  "refDistance",
+  "rolloffFactor",
+  "coneInnerAngle",
+  "coneOuterAngle",
+  "coneOuterGain"
+] as const;
+type ReducedAudioSettingsKeys = typeof REDUCED_KEYS[number];
+type ReducedAudioSettings = Pick<AudioSettings, ReducedAudioSettingsKeys>;
+
+// We apply the most restrictive audio parameters
+const paramsReducer = (acc: AudioSettings, curr: AudioSettings): AudioSettings => {
+  if (!curr && !acc) return {} as AudioSettings;
+  else if (curr && !acc) return curr;
+  else if (!curr && acc) return acc;
+  else
+    return REDUCED_KEYS.reduce((result: ReducedAudioSettings, key: ReducedAudioSettingsKeys): AudioSettings => {
+      if (curr[key] !== undefined && acc[key] !== undefined) {
+        result[key] = Math.min(acc[key]!, curr[key]!);
+      } else if (curr[key] !== undefined && acc[key] === undefined) {
+        result[key] = curr[key];
+      } else if (curr[key] === undefined && acc[key] !== undefined) {
+        result[key] = acc[key];
+      }
+      return result;
+    }, {});
+};
+
+const addOrRemoveZone = (zones: Set<number>, zone: number, position: Vector3) => {
+  if (!zones) return;
+  const isInZone = isZoneEnabled(zone) && zoneContains(zone, position);
+  const wasInZone = zones.has(zone);
+  if (isInZone && !wasInZone) {
+    zones.add(zone);
+  } else if (!isInZone && wasInZone) {
+    zones.delete(zone);
+  }
+};
+
+const currZones = new Map<number, Set<number>>();
+const prevZones = new Map<number, Set<number>>();
+const aabbs = new Map<number, Box3>();
+
+const listenerPos = new Vector3();
+const audioZoneQuery = defineQuery([AudioZone]);
+const audioZoneEnterQuery = enterQuery(audioZoneQuery);
+const audioZoneExitQuery = exitQuery(audioZoneQuery);
+const audioZoneSourceQuery = defineQuery([AudioEmitter]);
+const audioZoneSourceEnterQuery = enterQuery(audioZoneSourceQuery);
+const audioZoneSourceExitQuery = exitQuery(audioZoneSourceQuery);
+const audioZoneListenerQuery = defineQuery([AudioListenerTag]);
+const audioZoneListenerEnterQuery = enterQuery(audioZoneListenerQuery);
+const audioZoneListenerExitQuery = exitQuery(audioZoneListenerQuery);
+
+export function audioZoneSystem(world: HubsWorld) {
+  const zones = audioZoneQuery(world);
+  if (!zones.length) return;
+
+  [...audioZoneSourceEnterQuery(world), ...audioZoneListenerEnterQuery(world)].forEach(entity => {
+    currZones.set(entity, new Set());
+    prevZones.set(entity, new Set());
+  });
+  [...audioZoneSourceExitQuery(world), ...audioZoneListenerExitQuery(world)].forEach(entity => {
+    currZones.delete(entity);
+    prevZones.delete(entity);
+  });
```

Comment:
It feels strange that the audio listener needs to keep track of the zones that it is in. Isn't each zone keeping track of which entities are inside it? In this case, the zones that an audio listener is in each frame is a _derived property_. Even if we want to cache it to compare it each frame, it feels strange to store the information in the same maps.

In other words, `currZones` and `prevZones` are used to store two very different types of associations:
- On the one hand, they associate the `EntityID` of an `audioZone` to a `Set` of entities that are inside that zone. 
- On the other hand, they also associate the `EntityID` of an `audioListener` to a `Set` of `audioZone EntityIDs` that the listener is inside of.

Reply:
In current approach each entity keeps track of what zones it's in, so `currZones` and `prevZones` both have `entityEid->Set<zoneEid>` associations. Both the audio emitter and the audio listener are audio zone entities but we keep the listener as a special case because we handle things like the world pos calculation in a different way because of the different nature of the entity.

Conceptually both the `AudioListener` and the `AudioEmitter` components should also be `AudioZoneEntity` but it didn't make sense to add a tag when you can handle all of them in an homogenous way.

Reply:
Ah, right. My mistake. I misread `audioZoneSourceEnterQuery` as `audioZoneEnterQuery` and assumed it was `EntityID`s of `AudioZone`s